feat: stageSelection 버튼 기능 구현 및 모달 창 구현

yourchaee opened the pull request:
## 개요
- 게임 메뉴의 `reset` 버튼과 `stageSelection` 버튼 기능 구현
- 버튼 클릭 시 생성되는 모달 창 구현

## 작업 내용

### StageSelection 버튼 기능
- `stageSelection` 버튼을 클릭 시 stage 선택 화면으로 이동하도록 `redux`의 `useDispatch`, `setMode`를 활용했습니다.
```javascript
const dispatch = useDispatch();
const handleStageSelection = () => {
  dispatch(setMode("StageSelection"));
};
```

https://user-images.githubusercontent.com/110456453/226925389-0cf7a7e2-5fef-46a4-b8fb-c3b4e0887add.mov


### 모달 창
- Modal: 모달 공통 컴포넌트
- ModalWrapper : 모달을 감싸는 부분
- 두 부분으로 나누어 하나의 모달을 형성했습니다.
- ModalWrapper 부분을 클릭 시 모달 창이 닫히는 기능을 만들기 위해 부분을 나누었습니다.
- `useState`를 이용해 모달 창이 닫히는 기능을 구현했습니다.


https://user-images.githubusercontent.com/110456453/226925371-91ada6c4-5255-4be4-9ffc-f6bafe13daeb.mov


### `reset` 버튼 기능 
- `redux`의 `useSelector` 를 활용해 현재 스테이지 정보를 확인 후 해당 스테이지의 초기값으로 되돌리는 기능을 구현하려고 했습니다.
-`redux` 사용시 eslint `import/no-cycle` 에러가 발생합니다.
- 해당 에러는 두 컴포넌트 간 끝없는 infinite loop을 발생시킵니다.
- 위의 에러를 발생시키지 않는 방법은 스테이지 별 조명, 출발지점, 오브젝트 방향 등등 스테이지 구성을 완료한 후 해당 초기값으로 돌리는 reset 버튼을 구현하는 방법이며, 이 방법은 스테이지 구성이 끝난 후 진행하기로 보류했습니다.

## 특이 사항
N/A

yourchaee committed:
feat: 이벤트 모달 창 생성

- 리셋 버튼 클릭 시 모달 창 생성
- 스테이지 선택 버튼 클릭 시 모달 창 생성

yourchaee committed:
feat: 스테이지 선택 버튼 기능 구현

- 스테이지 버튼 클릭 시 모달 창 팝업
- 모달 창 Yes 버튼 클릭 시 스테이지 선택 화면으로 이동

yourchaee committed:
feat: 리셋 버튼 모달 창 생성

- 리셋 버튼 클릭 시 모달 창 생성
- 모달 창 주위 클릭 시 모달 창 없어지는 기능

yourchaee committed:
feat: 모달 창 닫는 기능

- 모달 창 주위를 클릭 시 모달창이 닫히는 기능

yourchaee committed:
feat: 리셋 버튼 클릭 후 모달 창 닫는 기능

- 리셋 버튼 클릭 후 모달 창 열리면, 주위 클릭 시 모달 창 닫히는 기능

yourchaee committed:
fix: modal 폴더 구조 및 컴포넌트 분리
- 스테이지 선택 버튼 컴포넌트 분리
- 리셋 버튼 컴포넌트 분리

sageherb commented on src/components/menus/GameMenu.js:
```diff
@@ -1,23 +1,29 @@
 
 export default function GameMenu() {
-  const dispatch = useDispatch();
+  const [reset, setReset] = useState(false);
+  const [selectStage, setSelectStage] = useState(false);
```
초기 상태 외에는 false가 되는 로직이 없는 것 같아요.
확인해보니 모달 닫고 다시 누르면 이 값이 계속 `true`라서 모달이 열리지 않습니다!

yourchaee replied:
넵 업로드 후에 버튼을 두번째 클릭 시 모달이 열리지 않는다는 걸 알았습니다.
수정해놓겠습니다!

sageherb commented on src/components/modals/StageSelectModal.js:
```diff
@@ -0,0 +1,50 @@
+  width: 100vw;
+  height: 100vh;
+  background-color: rgba(0, 0, 0, 0.7);
+`;
```
`ResetModal`과 `StageSelectModal`이 팝업 모달 안의 컴포넌트를 제외하곤 똑같은데, 하나의 컴포넌트로도 관리가 가능하지 않을까요?

yourchaee replied:
넵넵 모달 하나의 공통 컴포넌트로 합치겠습니다~

j25nkh reviewed (changes requested):
리셋과 스테이지 선택에 사용될 모달을 제작해 주셨네요.
Modal 컴포넌트 하나를 만들어 재사용이 가능해 보입니다. 마찬가지로 Button 컴포넌트도 재사용이 가능할 것 같아요.

j25nkh commented on src/components/modals/function/ResetFunction.js:
```diff
@@ -0,0 +1,12 @@
+
+import ResetButton from "../buttons/ResetButton";
+
+export default function ResetFunction() {
```
Component 이름에 Function이 적합하지 않은 것 같습니다.
더 나아가 별도의 컴포넌트를 만드는 이점이 그렇지 않은 경우보다 큰지 고민해보면 좋을 것 같습니다.

yourchaee replied:
덕분에 별도의 컴포넌트로 구분할 필요가 없어져서 해당 컴포넌트는 삭제했습니다~

j25nkh commented on src/components/modals/function/ResetFunction.js:
```diff
@@ -0,0 +1,12 @@
+  return (
+    <>
+      <h4>리셋 하시겠습니까?</h4>
+      <ResetButton />
```
현재는 Yes라고 레이블 되어있는 버튼이 하나만 있는데, UX를 감안하면 취소 또는 돌아가기 버튼을 추가하는게 어떨까요?

yourchaee replied:
의견 감사합니다! 취소 버튼을 추가했습니다 :)

yourchaee committed:
fix: PR 리뷰 반영

- 모달 창 버튼 클릭 시 열리는 기능 수정
- 모달 공통 컴포넌트로 수정
- 취소 버튼, yes 버튼 공통 컴포넌트로 수정
- 취소 버튼 추가
- 명확한 변수명 수정
- UI 수정
- 함수, props 네이밍 수정
- 모달 컴포넌트 메세지 수정

j25nkh commented on src/components/menus/GameMenu.js:
```diff
@@ -1,22 +1,31 @@
+    setIsOpen(true);
+    setMessage("스테이지를 선택하시겠습니까?");
+  };
+  const handleClose = () => {
```
handleModalClose 처럼 '닫아지는 주체'가 들어간 함수명은 어떨까요?

j25nkh commented on src/components/menus/GameMenu.js:
```diff
@@ -1,22 +1,31 @@
-
-  const handleStageSelection = () => {
-    dispatch(setMode("StageSelection"));
+  const [isOpen, setIsOpen] = useState(false);
```
열리는 주체가 어떤건지 상태명에 들어가면 좋을 것 같습니다 :)

j25nkh commented on src/components/modals/Modal.js:
```diff
@@ -0,0 +1,54 @@
+export default function Modal({ handleClose, message }) {
+  return (
+    <ModalWrapper>
+      <PopUpModal>
```
네이밍은 정답이 없으나, 검정색 배경부분은 ModalBackground,
내용과 버튼이 담겨있는 안쪽 부분은 ModalBox 어떠신가요?

yourchaee replied:
ModalArea로 변경해보았습니다..!

sageherb commented on src/components/menus/GameMenu.js:
```diff
@@ -1,22 +1,31 @@
+  const [message, setMessage] = useState("");
+  const handleResetClick = () => {
+    setIsOpen(true);
+    setMessage("리셋 하시겠습니까?");
```
안내 메시지는 게임 메뉴가 아니라 모달 쪽에 있는 게 더 자연스러울 것 같습니다!

yourchaee replied:
어떤 이유에서 모달쪽에 있는게 자연스러운지 궁금합니다.

sageherb reviewed (commented):
이벤트 핸들러 관련 이름이 전체적으로 좀 어색한 느낌을 받았습니다!
참고하실 링크 적어놓을게요
https://blog.sonim1.com/220
https://react.dev/learn/responding-to-events
https://youngmin.hashnode.dev/react-1

j25nkh reviewed (approved):
고생 많으셨습니다 :) 👍

sageherb reviewed (approved):
깔끔하게 해주셨네요! 고생하셨습니다 👏
